group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/AbstractUniformAssignor.java

```diff
@@ -0,0 +1,220 @@
+            return optimizedUniformAssignor.build();
+        } else {
+            GeneralUniformAssignor generalUniformAssignor = new GeneralUniformAssignor();
+            log.debug("Detected that all members are subscribed to a different set of topics, invoking the "
+                + "general assignment algorithm");
+            return generalUniformAssignor.build();
+        }
+    }
+
+    /**
+     * Builds the assignment for the given topic and member metadata.
+     *
+     * @return The new assignment for the group.
+     */
+    protected abstract GroupAssignment build();
+
+    private boolean allSubscriptionsEqual(Map<String, AssignmentMemberSpec> members) {
+        boolean areAllSubscriptionsEqual = true;
+        Collection<Uuid> firstSubscriptionList = members.values().iterator().next().subscribedTopicIds();
+        for (AssignmentMemberSpec memberSpec : members.values()) {
+            if (!firstSubscriptionList.equals(memberSpec.subscribedTopicIds())) {
+                areAllSubscriptionsEqual = false;
+                break;
+            }
+        }
+        return areAllSubscriptionsEqual;
+    }
+
+    protected boolean useRackAwareAssignment(Set<String> consumerRacks, Set<String> partitionRacks, Map<TopicIdPartition, Set<String>> racksPerPartition) {
+        if (consumerRacks.isEmpty() || Collections.disjoint(consumerRacks, partitionRacks))
+            return false;
+        else {
+            return !racksPerPartition.values().stream().allMatch(partitionRacks::equals);
+        }
+    }
+
+    protected List<TopicIdPartition> getAllTopicIdPartitions(List<Uuid> listAllTopics, SubscribedTopicDescriber subscribedTopicDescriber) {
+        List<TopicIdPartition> allPartitions = new ArrayList<>();
+        for (Uuid topic : listAllTopics) {
+            int partitionCount = subscribedTopicDescriber.numPartitions(topic);
+            for (int i = 0; i < partitionCount; ++i) {
+                allPartitions.add(new TopicIdPartition(topic, i));
+            }
+        }
+        return allPartitions;
+    }
+
+    protected class RackInfo {
+        protected final Map<String, String> memberRacks;
+        protected final Map<TopicIdPartition, Set<String>> partitionRacks;
+        private final Map<TopicIdPartition, Integer> numMembersByPartition;
+
+        public RackInfo(AssignmentSpec assignmentSpec, SubscribedTopicDescriber subscribedTopicDescriber, List<Uuid> topicIds) {
+            Map<String, List<String>> membersByRack = new HashMap<>();
+            assignmentSpec.members().forEach((memberId, assignmentMemberSpec) ->
+                assignmentMemberSpec.rackId().filter(r -> !r.isEmpty()).ifPresent(
+                    rackId -> membersByRack.computeIfAbsent(rackId, k -> new ArrayList<>()).add(memberId)
+                )
+            );
+
+            Set<String> allPartitionRacks;
```

Comment:
Collections.emptyMap is immutable though right, ig a better option would be to initliaze them as new HashMaps

Reply:
you can't change what's stored in an empty map but you can change which map the variable points to. though it seems fine in this case since we are using streams api and need a final variable